File: src/main/java/xyz/phanta/cbtweaker/integration/jei/JeiAccumulatorMap.java
package xyz.phanta.cbtweaker.integration.jei;

import xyz.phanta.cbtweaker.buffer.BufferType;

import javax.annotation.Nullable;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

public class JeiAccumulatorMap {

    private final Map<String, Group> groupTable;

    public JeiAccumulatorMap(Map<String, JeiBufferGroup> bufGroups) {
        this.groupTable = bufGroups.entrySet().stream()
                .collect(Collectors.toMap(Map.Entry::getKey, e -> new Group(e.getValue())));
    }

    @Nullable
    public Group getGroup(String groupId) {
        return groupTable.get(groupId);
    }

    public static class Group {

        private final Map<BufferType<?, ?, ?, ?>, Object> accumTable = new HashMap<>();

        public Group(JeiBufferGroup bufGroup) {
            bufGroup.forEach(new JeiBufferGroup.Visitor() {
                @Override
                public <JB, JA> void visit(BufferType<?, ?, JB, JA> bufType, List<JB> buffers) {
                    JA accum = bufType.createJeiAccumulator();
                    for (JB buffer : buffers) {
                        bufType.jeiAccumulate(accum, buffer);
                    }
                    accumTable.put(bufType, accum);
                }
            });
        }

        @SuppressWarnings("unchecked")
        @Nullable
        public <JA> JA getAccumulator(BufferType<?, ?, ?, JA> bufType) {
            return (JA)accumTable.get(bufType);
        }

    }

}

File: src/main/kotlin/st/evening/mc/cbtweaker/compat/jei/recipe/JeiAccumulatorMap.kt
package st.evening.mc.cbtweaker.compat.jei.recipe

import st.evening.mc.cbtweaker.buffer.BufferType

class JeiAccumulatorMap(bufGroups: Map<String, JeiBufferGroup>) {
    private val groupTable: Map<String, Group> = bufGroups.mapValues { Group(it.value) }

    operator fun contains(bufGroupId: String): Boolean = bufGroupId in groupTable

    operator fun get(bufGroupId: String): Group? = groupTable[bufGroupId]

    class Group(bufGroup: JeiBufferGroup) {
        private val accumTable: MutableMap<BufferType<*, *, *, *>, Any?> = mutableMapOf()

        init {
            bufGroup.forEach(object : JeiBufferGroup.Visitor {
                override fun <JB, JA> visit(bufType: BufferType<*, *, JB, JA>, buffers: Map<String, JB>) {
                    val acc = bufType.createJeiAccumulator()
                    buffers.values.forEach {
                        bufType.jeiAccumulate(acc, it)
                    }
                    accumTable[bufType] = acc
                }
            })
        }

        @Suppress("UNCHECKED_CAST")
        fun <JA> getAccumulator(bufType: BufferType<*, *, *, JA>): JA? = accumTable[bufType] as? JA
    }
}
